File: packages/zynth-network/ios/NetworkModels.swift
import Foundation

struct DiscoveryConfig {
  let serviceType: String
  let domain: String
  let resolveTimeoutMs: Int
}

struct AdvertiseOptions {
  let serviceType: String
  let name: String
  let domain: String
  let port: Int
  let txtRecord: [String: String]
}

struct AdvertisedServiceInfo {
  let serviceType: String
  let name: String
  let domain: String
  let port: Int
  let txtRecord: [String: String]

  func toDictionary() -> [String: Any] {
    return [
      "serviceType": serviceType,
      "name": name,
      "domain": domain,
      "port": port,
      "txtRecord": txtRecord,
    ]
  }
}

struct NetworkServiceInfo {
  let id: String
  let name: String
  let type: String
  let domain: String
  var hostName: String?
  var port: Int
  var addresses: [String]
  var txtRecord: [String: String]
  var lastSeenAt: Int64

  func toDictionary() -> [String: Any] {
    return [
      "id": id,
      "name": name,
      "type": type,
      "domain": domain,
      "hostName": hostName ?? NSNull(),
      "port": port,
      "addresses": addresses,
      "txtRecord": txtRecord,
      "lastSeenAt": lastSeenAt,
    ]
  }
}

struct DiscoveryEvent {
  let type: String
  let timestamp: Int64
  let service: NetworkServiceInfo

  func toDictionary() -> [String: Any] {
    return [
      "type": type,
      "timestamp": timestamp,
      "service": service.toDictionary(),
    ]
  }
}

File: packages/zynth-network/android/ZynthNetwork/src/main/java/dev/zynth/network/NetworkModels.kt
package dev.zynth.network

import org.json.JSONArray
import org.json.JSONObject

data class DiscoveryConfig(
    val serviceType: String,
    val domain: String,
    val resolveTimeoutMs: Int,
)

data class AdvertisedServiceInfo(
    val serviceType: String,
    val name: String,
    val domain: String,
    val port: Int,
    val txtRecord: Map<String, String>,
) {
    fun toJson(): JSONObject {
        val txtRecordJson = JSONObject()
        txtRecord.forEach { (key, value) ->
            txtRecordJson.put(key, value)
        }

        return JSONObject()
            .put("serviceType", serviceType)
            .put("name", name)
            .put("domain", domain)
            .put("port", port)
            .put("txtRecord", txtRecordJson)
    }
}

data class NetworkServiceInfo(
    val id: String,
    val name: String,
    val type: String,
    val domain: String,
    val hostName: String?,
    val port: Int,
    val addresses: List<String>,
    val txtRecord: Map<String, String>,
    val lastSeenAt: Long,
) {
    fun toJson(): JSONObject {
        val addressesJson = JSONArray()
        addresses.forEach { address ->
            addressesJson.put(address)
        }

        val txtRecordJson = JSONObject()
        txtRecord.forEach { (key, value) ->
            txtRecordJson.put(key, value)
        }

        return JSONObject()
            .put("id", id)
            .put("name", name)
            .put("type", type)
            .put("domain", domain)
            .put("hostName", hostName ?: JSONObject.NULL)
            .put("port", port)
            .put("addresses", addressesJson)
            .put("txtRecord", txtRecordJson)
            .put("lastSeenAt", lastSeenAt)
    }
}

data class DiscoveryEvent(
    val type: String,
    val timestamp: Long,
    val service: NetworkServiceInfo,
) {
    fun toJson(): JSONObject {
        return JSONObject()
            .put("type", type)
            .put("timestamp", timestamp)
            .put("service", service.toJson())
    }
}
